Add ability for user to manually override userAgentData

colelao opened the pull request:
#### 48b4c083dd541a67eba3a63adc8495144331f481
<pre>
Add ability for user to manually override userAgentData
<a href="https://bugs.webkit.org/show_bug.cgi?id=296982">https://bugs.webkit.org/show_bug.cgi?id=296982</a>
<a href="https://rdar.apple.com/157612451">rdar://157612451</a>

Reviewed by Brent Fulgham.

This patch checks for a manually selected user agent string and, if present, parses the string and populates the userAgentData data structure with the results.

* Source/WTF/wtf/ASCIICType.h:
(WTF::isASCIIGraphic):
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/page/Navigator.cpp:
(WebCore::Navigator::userAgentData const):
* Source/WebCore/page/NavigatorUAData.cpp:
(WebCore::NavigatorUAData::NavigatorUAData):
(WebCore::NavigatorUAData::create):
(WebCore::NavigatorUAData::brands const):
(WebCore::NavigatorUAData::mobile const):
(WebCore::NavigatorUAData::platform const):
* Source/WebCore/page/NavigatorUAData.h:
* Source/WebCore/page/UserAgentStringData.h: Copied from Source/WebCore/page/NavigatorUAData.h.
(WebCore::UserAgentStringData::create):
* Source/WebCore/page/UserAgentStringParser.cpp: Added.
(WebCore::UserAgentStringParser::UserAgentStringParser):
(WebCore::UserAgentStringParser::create):
(WebCore::UserAgentStringParser::parse):
(WebCore::UserAgentStringParser::consumeProduct):
(WebCore::UserAgentStringParser::consumeRWS):
(WebCore::UserAgentStringParser::consumeComment):
(WebCore::UserAgentStringParser::consumeToken):
(WebCore::UserAgentStringParser::consumeQuotedPair):
(WebCore::UserAgentStringParser::peek):
(WebCore::UserAgentStringParser::increment):
(WebCore::UserAgentStringParser::atEnd):
(WebCore::UserAgentStringParser::isDelimiter const):
(WebCore::UserAgentStringParser::isTchar const):
(WebCore::UserAgentStringParser::isCtext const):
(WebCore::UserAgentStringParser::isObsText const):
(WebCore::UserAgentStringParser::isVisibleChar const):
(WebCore::UserAgentStringParser::isRWS const):
* Source/WebCore/page/UserAgentStringParser.h: Copied from Source/WebCore/page/NavigatorUAData.h.

Canonical link: <a href="https://commits.webkit.org/298650@main">https://commits.webkit.org/298650@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/934f1da9b0b72147d51e58ec50cd68c5be0d9ea6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/116131 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/35792 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/26337 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/122187 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/66687 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/118020 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/36486 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/44380 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/88246 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/42768 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/119079 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/29111 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/104213 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/68657 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/28229 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/22322 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/65869 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/108241 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/98502 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/22457 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/125337 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/114659 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/43025 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/32309 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/96969 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/43390 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/100402 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/96753 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24625 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/42026 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/19916 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/38972 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/42912 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/48504 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/143356 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/42379 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/36954 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/45714 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/44083 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

colelao commented on Source/WebCore/page/ios/DOMTimerHoldingTank.h:
```diff
@@ -27,6 +27,7 @@
 
 #if ENABLE(CONTENT_CHANGE_OBSERVER)
 
+#include "DOMTimer.h"
```
This was for an include error due to unified sources being changed after adding new files.

brentfulgham commented on Source/WebCore/page/Navigator.cpp:
```diff
@@ -450,9 +454,22 @@ int Navigator::maxTouchPoints() const
+            auto userAgentString = client.userAgent(document->url());
+            Ref parser = UserAgentStringParser::create(userAgentString);
+            std::optional userAgentStringData = parser->parse();
+            if (userAgentStringData.has_value()) {
```
You can just write this as:
`if (userAgentStringData) {`

It will do a `bool` conversion on-the-fly.

brentfulgham commented on Source/WebCore/page/Navigator.cpp:
```diff
@@ -450,9 +454,22 @@ int Navigator::maxTouchPoints() const
-        m_navigatorUAData = NavigatorUAData::create();
+    RefPtr frame = this->frame();
+    if (frame && frame->page()) {
+        auto document = frame->document();
```
I think it might be simpler to use `Document::userAgent(const URL&)` here (with a blank URL). Actually, since you have the frame already, you can just ask it's loader: This will check if any quirks are enabled that change UA and return the correct thing, check for custom UserAgent and return that if relevant, and finally return the standard UA if none of these special cases matter. (Look at `FrameLoader::userAgent(const URL&)` for details of the logic).

Then you can just make this code:
```
if (frame && frame->page()) {
    auto userAgentString  = frame->loader().userAgent({});
    Ref parser = UserAgentStringParser::create(userAgentString);
    std::optional userAgentStringData = parser->parse();
    if (userAgentStringData) {
        ....
```

brentfulgham replied:
I think this might also future-proof this so that if we have multiple frames, with some needing different UA, script running in those frames will get the right thing (as opposed to just what the client suggested prior to considering quirks).

brentfulgham commented on Source/WebCore/page/Navigator.cpp:
```diff
@@ -450,9 +454,22 @@ int Navigator::maxTouchPoints() const
+            Ref parser = UserAgentStringParser::create(userAgentString);
+            std::optional userAgentStringData = parser->parse();
+            if (userAgentStringData.has_value()) {
+                m_navigatorUAData = NavigatorUAData::create(userAgentStringData.value());
```
Seems like you could just `WTFMove(*userAgentStringData)`. It's never used again after this point.

brentfulgham commented on Source/WebCore/page/UserAgentStringParser.h:
```diff
@@ -0,0 +1,75 @@
+    bool malformed = false;
+    const String& m_userAgentString;
+    size_t pos = 0;
+    size_t start = 0;
```
WebKit style is to use the brace initialization. This allows us to also include a bit count when packing:

`bool malformed { false };`
`size_t pos { 0 };`

brentfulgham commented on Source/WebCore/page/WorkerNavigator.cpp:
```diff
@@ -112,9 +114,14 @@ void WorkerNavigator::clearAppBadge(Ref<DeferredPromise>&& promise)
 {
-    if (!m_navigatorUAData)
-        m_navigatorUAData = NavigatorUAData::create();
+    Ref parser = UserAgentStringParser::create(m_userAgent);
```
This makes me wonder if workers don't pay attention to quirked user agents. Maybe @youennf knows.

brentfulgham commented on Source/WebCore/page/WorkerNavigator.cpp:
```diff
@@ -112,9 +114,14 @@ void WorkerNavigator::clearAppBadge(Ref<DeferredPromise>&& promise)
-        m_navigatorUAData = NavigatorUAData::create();
+    Ref parser = UserAgentStringParser::create(m_userAgent);
+    std::optional userAgentStringData = parser->parse();
+    if (userAgentStringData.has_value()) {
```
`if (userAgentStringData) {`

brentfulgham commented on Source/WebCore/page/WorkerNavigator.cpp:
```diff
@@ -112,9 +114,14 @@ void WorkerNavigator::clearAppBadge(Ref<DeferredPromise>&& promise)
+    Ref parser = UserAgentStringParser::create(m_userAgent);
+    std::optional userAgentStringData = parser->parse();
+    if (userAgentStringData.has_value()) {
+        m_navigatorUAData = NavigatorUAData::create(userAgentStringData.value());
```
use `WTFMove(*userAgentStringData)`

brentfulgham commented on Source/WebCore/page/UserAgentStringParser.cpp:
```diff
@@ -0,0 +1,347 @@
+
+namespace WebCore {
+UserAgentStringParser::UserAgentStringParser(const String& userAgentStringParser)
+    : m_userAgentString(userAgentStringParser), data(UserAgentStringData::create()) { };
```
The name of the `userAgentStringParser` argument seems wrong. We prefer to put each member initializer on its own line:

```
UserAgentStringParser::UserAgentStringParser(const String& userAgentString)
    : m_userAgentString(userAgentString)
    , data(UserAgentStringData::create())
 {
 }
```

brentfulgham commented on Source/WebCore/page/UserAgentStringParser.cpp:
```diff
@@ -0,0 +1,347 @@
+
+struct BrowsersSeen {
+    bool brave;
+    std::optional<String> braveVersion;
```
Can't we tell if `BrowserSeen` is true for each of these if we populate the Optional? If it doesn't have a value, the bool is false. If it does, the bool would have been set to true.  We can just do a test against optional<> as a boolean (or it's `has_value()` predicate).

colelao replied:
We could, but there's no guarantee that each product has a product-version. We could set the optional version to an empty string when we encounter a product with no version, but I thought that was less obvious.

brentfulgham commented on Source/WebCore/page/UserAgentStringParser.h:
```diff
@@ -0,0 +1,75 @@
+public:
+    static Ref<UserAgentStringParser> create(const String& userAgentString);
+    std::optional<Ref<UserAgentStringData>> parse();
+    UserAgentStringParser(const String& userAgentStringParser);
```
You should make this constructor private, since it's only used by the `create` function.

brentfulgham commented on Source/WebCore/page/NavigatorUAData.h:
```diff
@@ -50,7 +52,13 @@ class NavigatorUAData : public RefCounted<NavigatorUAData> {
 
 private:
     NavigatorUAData();
+    NavigatorUAData(Ref<UserAgentStringData>);
```
This should take a `Ref<UserAgentStringData>&&` since we always hand off ownership to this constructor.

colelao replied:
This is great to know

brentfulgham commented on Source/WebCore/page/NavigatorUAData.h:
```diff
@@ -35,10 +35,12 @@ namespace WebCore {
 class NavigatorUAData : public RefCounted<NavigatorUAData> {
 public:
     static Ref<NavigatorUAData> create();
+    static Ref<NavigatorUAData> create(Ref<UserAgentStringData>);
```
This should take a Ref<UserAgentStringData>&& since we always hand off ownership to the constructor.

brentfulgham commented on Source/WebCore/page/NavigatorUAData.cpp:
```diff
@@ -54,13 +57,35 @@ namespace WebCore {
 const Vector<NavigatorUABrandVersion>& NavigatorUAData::brands() const
 {
+    if (overrideFromUserAgentString)
+        return brandsOverride;
```
Do we need to add the randomized "Arbitrary UA Brand" to the vector we return?

brentfulgham commented on Source/WebCore/page/NavigatorUAData.h:
```diff
@@ -50,7 +52,13 @@ class NavigatorUAData : public RefCounted<NavigatorUAData> {
+    bool overrideFromUserAgentString = false;
+    bool mobileOverride = false;
+    Vector<NavigatorUABrandVersion> brandsOverride;
+    String platformOverride;
```
It might be a little cleaner if you stored each of these overrides in an optional<>. Then you could get rid of the separate `overrideFromUserAgentString` flag, since these would only ever have values if an override was in use.

brentfulgham commented on Source/WebCore/page/NavigatorUAData.cpp:
```diff
@@ -54,13 +57,35 @@ namespace WebCore {
 
+NavigatorUAData::NavigatorUAData(Ref<UserAgentStringData> userAgentStringData)
+{
+    overrideFromUserAgentString = true;
```
If you made `mobileOverride`, `brandsOverride`, and `platformOverride` into optional<>, then you could get rid of the separate flag.

brentfulgham commented on Source/WebCore/page/NavigatorUAData.cpp:
```diff
@@ -54,13 +57,35 @@ namespace WebCore {
+
 const Vector<NavigatorUABrandVersion>& NavigatorUAData::brands() const
 {
+    if (overrideFromUserAgentString)
```
... and then this could become:

```
if (brandsOverride)
    return *brandsOverride;
```

brentfulgham commented on Source/WebCore/page/NavigatorUAData.cpp:
```diff
@@ -54,13 +57,35 @@ namespace WebCore {
+        NavigatorUABrandVersion {
+            .brand = userAgentStringData->browserName,
+            .version = userAgentStringData->browserVersion
+        }
```
Maybe you need to do the `std::call_once` thing with the randomizer to set the "Arbitrary Browser" stuff in this array? So you might want to factor that into a helper function.

brentfulgham commented on Source/WebCore/page/NavigatorUAData.cpp:
```diff
@@ -85,6 +110,9 @@ const Vector<NavigatorUABrandVersion>& NavigatorUAData::brands() const
 bool NavigatorUAData::mobile() const
 {
+    if (overrideFromUserAgentString)
+        return mobileOverride;
```
And this would become:
```
if (mobileOverride)
    return *mobileOverride;
```

brentfulgham commented on Source/WebCore/page/NavigatorUAData.cpp:
```diff
@@ -96,6 +124,9 @@ bool NavigatorUAData::mobile() const
 String NavigatorUAData::platform() const
 {
+    if (overrideFromUserAgentString)
+        return platformOverride;
```
And this would become:
```
if (platformOverride)
    return *platformOverride;
```

brentfulgham commented on Source/WebCore/page/NavigatorUAData.cpp:
```diff
@@ -125,6 +156,12 @@ UALowEntropyJSON NavigatorUAData::toJSON() const
 void NavigatorUAData::getHighEntropyValues(const Vector<String>& hints, NavigatorUAData::ValuesPromise&& promise) const
 {
     auto values = UADataValues::create(brands(), mobile(), platform());
+    if (overrideFromUserAgentString) {
```
Oh! But what do you do here. I guess you do need the flag.

brentfulgham reviewed (changes requested):
This looks great! I think it might be better to use the FrameLoader approach I suggest, just because it handles a bunch of edge cases for weird UA swaps.

brentfulgham commented on Source/WebCore/page/Navigator.cpp:
```diff
@@ -450,9 +454,21 @@ int Navigator::maxTouchPoints() const
+    RefPtr frame = this->frame();
+    if (frame && frame->page()) {
+        RefPtr client = frame->loader().client();
+        if (client->hasCustomUserAgent()) {
```
Note: we should see if there are edge cases where the client is unaware of a custom user agent (e.g., a site specific quirk)

brentfulgham reviewed (approved):
Thank you for cleaning things up. Looks good!

colelao committed:
Add ability for user to manually override userAgentData
https://bugs.webkit.org/show_bug.cgi?id=296982
rdar://157612451

Reviewed by Brent Fulgham.

This patch checks for a manually selected user agent string and, if present, parses the string and populates the userAgentData data structure with the results.

* Source/WTF/wtf/ASCIICType.h:
(WTF::isASCIIGraphic):
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/page/Navigator.cpp:
(WebCore::Navigator::userAgentData const):
* Source/WebCore/page/NavigatorUAData.cpp:
(WebCore::NavigatorUAData::NavigatorUAData):
(WebCore::NavigatorUAData::create):
(WebCore::NavigatorUAData::brands const):
(WebCore::NavigatorUAData::mobile const):
(WebCore::NavigatorUAData::platform const):
* Source/WebCore/page/NavigatorUAData.h:
* Source/WebCore/page/UserAgentStringData.h: Copied from Source/WebCore/page/NavigatorUAData.h.
(WebCore::UserAgentStringData::create):
* Source/WebCore/page/UserAgentStringParser.cpp: Added.
(WebCore::UserAgentStringParser::UserAgentStringParser):
(WebCore::UserAgentStringParser::create):
(WebCore::UserAgentStringParser::parse):
(WebCore::UserAgentStringParser::consumeProduct):
(WebCore::UserAgentStringParser::consumeRWS):
(WebCore::UserAgentStringParser::consumeComment):
(WebCore::UserAgentStringParser::consumeToken):
(WebCore::UserAgentStringParser::consumeQuotedPair):
(WebCore::UserAgentStringParser::peek):
(WebCore::UserAgentStringParser::increment):
(WebCore::UserAgentStringParser::atEnd):
(WebCore::UserAgentStringParser::isDelimiter const):
(WebCore::UserAgentStringParser::isTchar const):
(WebCore::UserAgentStringParser::isCtext const):
(WebCore::UserAgentStringParser::isObsText const):
(WebCore::UserAgentStringParser::isVisibleChar const):
(WebCore::UserAgentStringParser::isRWS const):
* Source/WebCore/page/UserAgentStringParser.h: Copied from Source/WebCore/page/NavigatorUAData.h.

Canonical link: https://commits.webkit.org/298650@main
